python/src/opendp/_convert.py

```diff
@@ -204,38 +201,60 @@ def _slice_to_string(raw: FfiSlicePtr) -> str:
 
-def _vector_to_slice(val: Sequence[Any], type_name) -> FfiSlicePtr:
-    assert type_name[:4] == 'Vec<'
-    inner_type_name = type_name[4:-1]
+def _vector_to_slice(val: Sequence[Any], type_name: RuntimeType) -> FfiSlicePtr:
+    assert type_name.origin == 'Vec'
+    assert len(type_name.args) == 1, "Vec only has one generic argument"
+    inner_type_name = type_name.args[0]
 
+    # input is numpy array
+    # TODO: can we use the underlying buffer directly?
     if np is not None and isinstance(val, np.ndarray):
         val = val.tolist()
 
     if not isinstance(val, list):
-        raise OpenDPException(f"Cannot cast a non-list type to a vector")
+        raise TypeError(f"Cannot cast a non-list type to a vector")
 
     if inner_type_name == "String":
         def str_to_slice(val):
             return ctypes.c_char_p(val.encode())
         array = (ctypes.c_char_p * len(val))(*map(str_to_slice, val))
         return _wrap_in_slice(array, len(val))
 
+    if isinstance(inner_type_name, RuntimeType):
+        c_repr = [py_to_c(v, c_type=AnyObjectPtr, type_name=inner_type_name) for v in val]
+        array = (AnyObjectPtr * len(val))(*c_repr)
+        ffislice = _wrap_in_slice(array, len(val))
+        ffislice.depends_on(*c_repr)
+        return ffislice
+
     if inner_type_name not in ATOM_MAP:
-        raise OpenDPException(f"Members must be one of {ATOM_MAP.keys()}. Found {inner_type_name}.")
+        raise TypeError(f"Members must be one of {tuple(ATOM_MAP.keys())}. Found {inner_type_name}.")
 
     if val:
         # check that actual type can be represented by the inner_type_name
         equivalence_class = ATOM_EQUIVALENCE_CLASSES[str(RuntimeType.infer(val[0]))]
         if inner_type_name not in equivalence_class:
-            raise OpenDPException("Data cannot be represented by the suggested type_name")
+            raise TypeError("Data cannot be represented by the suggested type_name")
 
     array = (ATOM_MAP[inner_type_name] * len(val))(*val)
     return _wrap_in_slice(array, len(val))
 
 
-def _slice_to_vector(raw: FfiSlicePtr, type_name: str) -> List[Any]:
-    assert type_name[:3] == 'Vec'
-    inner_type_name = type_name[4:-1]
+def _slice_to_vector(raw: FfiSlicePtr, type_name: RuntimeType) -> List[Any]:
+    assert type_name.origin == 'Vec'
+    assert len(type_name.args) == 1, "Vec only has one generic argument"
+    inner_type_name = type_name.args[0]
+
+    if inner_type_name == 'AnyObject':
+        from opendp._data import ffislice_of_anyobjectptrs
+        raw = ffislice_of_anyobjectptrs(raw)
+        array = ctypes.cast(raw.contents.ptr, ctypes.POINTER(AnyObjectPtr))[0:raw.contents.len]
+        res = list(map(c_to_py, array))
+        # when the top-level AnyObject is freed, it recursively frees all anyobjects inside of it
+        # adjust the type of constituent AnyObjects so that __delete__ is not called when they are dropped
+        for elem in array:
+            elem.__class__ = ctypes.POINTER(AnyObject)
```

Comment:
This is a neat trick, we just need to remember the implications in potential future changes. I don't think it's possible now, but if there ever were a situation where Python could remove an AnyObject from its parent Vec, then we'd have to rethink this.